sklearn_numba_dpex/kmeans/drivers.py

```diff
@@ -0,0 +1,459 @@
+        The width of the window over the centroids is defined as
+        `centroids_window_width_multiplier x preferred_work_group_size_multiple`. The
+        higher the value, the higher the cost in shared memory. If None, will default
+        to 1. It is required to be a power of two.
+
+    centroids_window_height : int
+        The height of the window, counted as a number of features. The higher the
+        value, the higher the cost in shared memory. If None, will default to 16. It is
+        required to be a power of two.
+
+    global_mem_cache_size : int
+        Size in bytes of the size of the global memory cache size. If None, the value
+        will be automatically fetched with dpctl. It is used to estimate the maximum
+        number of copies of the array of centroids that can be used for privatization.
+
+    centroids_private_copies_max_cache_occupancy : float
+        A maximum fraction of global_mem_cache_size that is allowed to be expected for
+        use when estimating the maximum number of copies that can be used for
+        privatization. If None, will default to 0.7 .
+
+    device: str
+        A valid sycl device filter.
+
+    X_layout: str
+        'F' or 'C'
+
+    dtype: np.float32 or np.float64
+        The floating point precision that the kernels should use. If None, will adapt
+        to the dtype of the data, else, will cast the data to the appropriate dtype.
+
+    """
+
+    def __init__(
+        self,
+        preferred_work_group_size_multiple=None,
+        work_group_size_multiplier=None,
+        centroids_window_width_multiplier=None,
+        centroids_window_height=None,
+        global_mem_cache_size=None,
+        centroids_private_copies_max_cache_occupancy=None,
+        device=None,
+        X_layout=None,
+        dtype=None,
+    ):
+        dpctl_device = dpctl.SyclDevice(device)
+        device_params = _DeviceParams(dpctl_device)
+
+        self.global_mem_cache_size = (
+            global_mem_cache_size or device_params.global_mem_cache_size
+        )
+
+        self.preferred_work_group_size_multiple = _check_power_of_2(
+            preferred_work_group_size_multiple
+            or device_params.preferred_work_group_size_multiple
+        )
+
+        self.work_group_size_multiplier = _check_power_of_2(
+            work_group_size_multiplier
+            or (
+                device_params.max_work_group_size
+                // self.preferred_work_group_size_multiple
+            )
+        )
+
+        self.centroids_window_width_multiplier = _check_power_of_2(
+            centroids_window_width_multiplier or 1
+        )
+
+        self.centroids_window_height = _check_power_of_2(centroids_window_height or 16)
+
+        self.centroids_private_copies_max_cache_occupancy = (
+            centroids_private_copies_max_cache_occupancy or 0.7
+        )
+
+        self.device = dpctl_device
+
+        self.X_layout = X_layout or "F"
```

Comment:
I think we should add a hint on this: is this default based on any preference?
```suggestion
        # F-contiguous array are more suitable for this algorithm. 
        self.X_layout = X_layout or "F"
```

Reply:
My understanding of the good practices for kernel execution is that "F" will be faster. Also, in fact "C" does not work currently. It is planned to fix "C" and benchmark it. In the end we'll choose whichever happens to be the fastest in the benchmark but it's not clear yet.

Reply:
Please add a `TODO` inline comment to make this explicit.